pmlc/dialect/pxa/transforms/stencil_generic.cc

```diff
@@ -0,0 +1,265 @@
+  unsigned v1o = v1.ord();
+  v1.setOrd(v2.ord());
+  v2.setOrd(v1o);
+  std::swap(*v1, *v2);
+}
+
+} // namespace
+
+int64_t StencilGeneric::getIdxRange(mlir::BlockArgument idx) {
+  assert(idx.getArgNumber() < ranges.size());
+  return ranges[idx.getArgNumber()];
+}
+
+mlir::Optional<mlir::StrideInfo>
+StencilGeneric::getStrideInfo(mlir::Operation *op) {
+  auto cached = strideInfoCache.find(op);
+  if (cached != strideInfoCache.end()) {
+    return cached->second;
+  }
+  auto loadOp = llvm::dyn_cast<mlir::AffineLoadOp>(*op);
+  if (loadOp) {
+    auto strideInfo = computeStrideInfo(loadOp);
+    if (strideInfo.hasValue())
+      strideInfoCache.emplace(std::make_pair(op, strideInfo.getValue()));
+    return strideInfo;
+  }
+  auto storeOp = llvm::dyn_cast<mlir::AffineStoreOp>(*op);
+  if (storeOp) {
+    auto strideInfo = computeStrideInfo(storeOp);
+    if (strideInfo.hasValue())
+      strideInfoCache.emplace(std::make_pair(op, strideInfo.getValue()));
+    return strideInfo;
+  }
+  auto reduceOp = llvm::dyn_cast<AffineReduceOp>(*op);
+  if (reduceOp) {
+    auto strideInfo = computeStrideInfo(reduceOp);
+    if (strideInfo.hasValue())
+      strideInfoCache.emplace(std::make_pair(op, strideInfo.getValue()));
+    return strideInfo;
+  }
+  return llvm::None;
+}
+
+void StencilGeneric::BindIndexes(
+    const llvm::SmallVector<mlir::Operation *, 3> &tensors) {
+  llvm::SmallVector<mlir::BlockArgument, 8> emptyBoundIdxsVector;
+  RecursiveBindIndex(&emptyBoundIdxsVector, tensors);
+}
+
+// TODO: Better to maintain boundIdxs as both set & vector, or just vector?
+//   Or could maintain as map from BlockArg to numeric index (i.e. where it
+//   would have been were it a vector)
+// TODO: Also, should probably at least be a small vector (how small?)
+void StencilGeneric::RecursiveBindIndex(
+    llvm::SmallVector<mlir::BlockArgument, 8> *boundIdxs,
+    const llvm::SmallVector<mlir::Operation *, 3> &tensors) {
+  auto currIdx = boundIdxs->size();
+  if (currIdx == semanticIdxCount) {
+    // This is a legal binding, go find a tiling for it
+    llvm::SmallVector<int64_t, 8> currTileSize(semanticIdxCount);
+    RecursiveTileIndex(TensorAndIndexPermutation(tensors, *boundIdxs),
+                       &currTileSize, 0);
+  } else {
+    for (const auto &blockArg : blockArgs) {
+      // Don't bind same index twice
+      if (std::find(boundIdxs->begin(), boundIdxs->end(), blockArg) !=
+          boundIdxs->end()) {
+        continue;
+      }
+
+      // Verify the requirements for this index with each tensor are all met
+      bool reqsMet = true;
+      for (unsigned i = 0; i < tensors.size(); i++) {
+        try { // TODO: probably don't keep long term
+          if (!requirements.at(std::make_pair(i, currIdx))(tensors[i],
+                                                           blockArg)) {
+            reqsMet = false;
+            break;
+          }
+        } catch (const std::out_of_range &e) {
```

Comment:
Is this throw caused my requirements.at(....)?  Seems incorrect handling.  I'd imagine if we don't have any specific requirements for a tensor/index pair, that would mean that any mapping is OK, so we should probably do:

auto it = requriements.find(...)
if (it != requirements.end() && !it->second(tensors[i], blockArg) {
  reqsMet = false;
  break;
}

Reply:
:+1: yes I agree that seems like the right way to handle this